grafana-operator/grafana-operator
# support folder-permissions

## Description
In Grafana, Dashboard-permissions cannot be more restrictive, than the folders they're in.
Thus it's necessary, to be able to configure the folder-permissions.

I added a new type for folders (together with the obligatory controller, client and config) to do exactly that.

**notes / known issues:**
* In Grafana the folder-title is not unique (it relies on the uid). As the folders are referenced from the GrafanaDashboards via customFolder**Name**, the `uid` will be reproducible generated from the title, making it unique as well.
* ATM the folders won't be deleted by the Reconciler (because we don't need that currently, and I'm not sure, if it might cause problems in certain circumstances ...and it can be implemented later anyway)
* Support to set folder-permissions for teams or users is not yet implemented - only roles are supported ATM (but should be easy to implement later - it's just, that the json-payload requires a different type for it)


_Bugfix:_
Along the way, I added more log-output and in `main()` I moved `printVersion()` to be executed _after_ the logger has been set/initialized. Its output was missing from the k8s-log otherwise.

## Type of change


## Checklist


I added a few unit-tests, but not enough to claim, that it proves the effectiveness.

## Verification steps


* deploy the grafana-operator with the changes (don't forget to apply the new CRDs!)
* add the k8s-config for the new kind `GrafanaFolder` to your configuration.
* Configure the permissions to not include Viewers (There is an example how this should look like in: `deploy/examples/folders/restricted-folder.yaml`)
* _optional_: put a `GrafanaDashboard` in the folder by referencing its title in `customFolderName`
* deploy/apply it
* open a web-browser of your choice and open the grafana-instance and navigate to "dashboards/manage" (or add `/dashboards` in the url-bar)
* Verify that a folder with the configured title exists
* hover over it and click on "Go to folder"
* click on the tab "Permissions"
* You should only see entries for "Admin" and "Editor" - compare with the default-folder (should have the title of the k8s-namespace its in), which also has "Viewer" configured
